### DISCUSSION TITLE
Can KeycloakJS library automatically logout a user, if their session gets invalidated via the admin panel?

### DISCUSSION THREAD

**Discussion by IgorMaj** [keycloak/keycloak#26279] [Q&A] [answered]
### Description

The title says most of it. Right now the user is technically logged out, but a page refresh is needed to trigger a redirect to the keycloak login page. 

Can Keycloak JS library catch this logout event, via some polling or in some other (more efficient) way?  My application is a SPA react application with the standard authentication flow. 

Thanks for any help and suggestions.

**Comment by jonkoops** [comment, marked as answer]
Yes, this is possible. By default Keycloak JS will call the `onAuthLogout` function when a user is signed out in another place. In the background this works through the [Session Status Iframe](https://www.keycloak.org/docs/latest/securing_apps/#session-status-iframe), which implements [OpenID Connect Session Management 1.0](https://openid.net/specs/openid-connect-session-1_0.html). You can access this event as follows:

```js
const keycloak = new Keycloak();

keycloak.onAuthLogout = () => {
  console.log("User logged out");
};
```

Note though that a lot of modern browsers have [tracking protection](https://www.keycloak.org/docs/latest/securing_apps/#_modern_browsers) that prevent this functionality from working as expected. We're hoping that we can implement the Federated Credential Management API (#16834) to resolve this issue in the future.

**Reply by IgorMaj** [reply]
Thanks for the quick reply!  I've already tried `onAuthLogout`, however I didn't take  tracking protection into consideration.

**Reply by jonkoops** [reply]
Yeah, it's likely the tracking protection, newer versions of Keycloak JS will give you a warning in the browser console if it detects this functionality is being blocked. Unfortunately with Google now rolling out their cookie protection this year I don't think there are any browsers left that support this.

The exception would be if your application is also hosted on the same domain as your Keycloak server, as that would not trigger the third-party cookies.
